diff --git a/tango/test_context.py b/tango/test_context.py
index 5f271e1b..d96f5b23 100644
--- a/tango/test_context.py
+++ b/tango/test_context.py
@@ -47,6 +47,9 @@ __all__ = (
 
 _DEFAULT_THREAD_TIMEOUT = 5.0
 _DEFAULT_PROCESS_TIMEOUT = 7.0
+_DEFAULT_GET_SERVER_PORT_RETRIES = 400
+_DEFAULT_GET_SERVER_PORT_DELAY = 0.03
+
 
 IOR = collections.namedtuple(
     "IOR",
@@ -82,7 +85,7 @@ def get_server_host_port():
     return ior.host.decode(), ior.port
 
 
-def get_server_port_via_pid(pid, host, retries=400, delay=0.03):
+def get_server_port_via_pid(pid, host, retries=_DEFAULT_GET_SERVER_PORT_RETRIES, delay=_DEFAULT_GET_SERVER_PORT_DELAY):
     """Return the TCP port that a device server process is listening on (GIOP).
 
     This checks TCP sockets open on the process with the given PID, and attempts
@@ -516,7 +519,8 @@ class MultiDeviceTestContext:
             if not self.port:
                 util = Util.instance()
                 pid = util.get_pid()
-                self.port = get_server_port_via_pid(pid, self.host)
+                retries = int(self.timeout * 0.75 / _DEFAULT_GET_SERVER_PORT_DELAY)
+                self.port = get_server_port_via_pid(pid, self.host, retries=retries)
             if self._process:
                 # set TANGO_HOST override for device server (DS) launcher child process
                 self._override_test_context_tango_host()
@@ -585,7 +589,7 @@ class MultiDeviceTestContext:
             db.put_device_attribute_property(device_name, properties_to_save)
         try:
             validated_db = Database(self.db)
-        except DevFailed:
+        except DevFailed as exc:
             with open(self.db) as f:
                 content = f.read()
             raise RuntimeError(
@@ -593,7 +597,7 @@ class MultiDeviceTestContext:
                 f"Check device properties (empty list or str?): "
                 f"{device_prop_info}.\n"
                 f"Problematic file has content:\n{content}"
-            ) from None
+            ) from exc
         return validated_db
 
     def delete_db(self):
@@ -639,20 +643,20 @@ class MultiDeviceTestContext:
         try:
             self._wait_until_port_is_known()
             self._wait_until_startup_status_is_known()
-        except RuntimeError:
+        except RuntimeError as exc:
             if self.thread.is_alive():
                 raise RuntimeError(
                     "The server appears to be stuck at initialization. Check stdout/stderr for more information."
-                ) from None
+                ) from exc
             elif hasattr(self.thread, "exitcode"):
                 raise RuntimeError(
                     f"The server process stopped with exitcode {self.thread.exitcode}. "
                     f"Check stdout/stderr for more information."
-                ) from None
+                ) from exc
             else:
                 raise RuntimeError(
                     "The server stopped without reporting. Check stdout/stderr for more information."
-                ) from None
+                ) from exc
 
         if self._startup_exception:
             raise self._startup_exception
